Create Authentication module

gleybersonandrade opened the pull request.

cmagnobarbosa committed:
Decorator @Authenticated works.

cmagnobarbosa committed:
Added Authenticated decorator

gleybersonandrade committed:
Update Auth module and include Authenticated decorator

cmagnobarbosa committed:
Auth login endpoint updated.

cmagnobarbosa committed:
Linter fix

ajoaoff reviewed (changes requested):
The methods to manipulate the user database should themselves be used only by authenticated users. There should be a method to create an initial user on deploy, or through CLI on the server.

gleybersonandrade committed:
Secure default authentication endpoints

Endpoints to create, update and delete user are secure now.

ajoaoff reviewed (changes requested):
I believe `_list_users` and `_list_user` should also require an authenticated user.
The method to create the default user should not be called by `__init__` , because the default user will be recreated every time Kytos is started.

gleybersonandrade committed:
Update conf template to include jwt secret

gleybersonandrade committed:
Create parameter to create superuser

gleybersonandrade committed:
Secure list user and list users endpoints

Endpoints to list users are secure now, and password no longer displayed

gleybersonandrade committed:
The superuser can be created using the -C parameter.

gleybersonandrade committed:
Create unit test for Auth module

cmagnobarbosa committed:
Added exception to get error on login in authentication module

An exception has been added at Login to get AttributeError when an unregistered user tries to login.

cmagnobarbosa committed:
Improved error exception message

gleybersonandrade committed:
Update unit test

The auth unit test is deprecated. This commit update it.

cmagnobarbosa committed:
Updated list_user and update_user in authentication unit test

This commit update unit tests to list the user and update a user.

ajoaoff commented on kytos/core/auth.py:
```diff
@@ -0,0 +1,334 @@
+                response = {
+                    "answer": {"data": box.data},
+                    "code": HTTPStatus.OK.value,
+                }
```
An empty `box` is not an error, it means there is no user with the given uid.

ajoaoff commented on kytos/core/auth.py:
```diff
@@ -0,0 +1,334 @@
+
+        def _find_user_callback(_event, box, error):
+            nonlocal response
+            if error or not box:
```
```suggestion
            if error:
```

ajoaoff commented on kytos/core/auth.py:
```diff
@@ -0,0 +1,334 @@
+                    "answer": "User data cannot be shown",
+                    "code": HTTPStatus.INTERNAL_SERVER_ERROR.value,
+                }
+            else:
```
```suggestion
            elif not box:
                response = {
                    "answer": f'User with uid {uid} not found',
                    "code": HTTPStatus.NOT_FOUND.value
                }
            else:
```

ajoaoff commented on kytos/core/auth.py:
```diff
@@ -0,0 +1,334 @@
+                response = {
+                    "answer": "User successfully deleted",
+                    "code": HTTPStatus.OK.value,
+                }
```
Same as `_find_user`. Here maybe it would be better to check first if the user exists.

gleybersonandrade committed:
Update kytos/core/auth.py

Co-Authored-By: Antonio Francisco <antonio@ansp.br>
Update kytos/core/auth.py

Co-Authored-By: Antonio Francisco <antonio@ansp.br>

hdiogenes commented on kytos/core/config.py:
```diff
@@ -76,6 +76,10 @@ def __init__(self):
                             action='store_true',
                             help="Enable all new Entities by default.")
 
+        parser.add_argument('-C', '--create_superuser',
```
```suggestion
        parser.add_argument('-C', '--create-superuser',
```
